<a href="https://colab.research.google.com/github/wallik2/SCIM_Lecture/blob/main/Intro_Data_Sci/Week2_Pandas_Application.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Pandas Application

In [29]:
import numpy as np
import pandas as pd

## <u>Application 1</u> : Compute the Student's Grade




### 1.1 ) Create the DataFrame

In [30]:
DF = pd.DataFrame(
    [[62, 'C'],[77, 'B+'],[27, 'F'],[44, 'F'],[58, 'D+'],[76, 'B+'],[30, 'F'],
     [53, 'D'],[20, 'F'],[76, 'B+'],[38, 'F'],[71, 'B'],[48, 'F'],[37, 'F'],
     [52, 'D'],[57, 'D+'],[69, 'C+'],[47, 'F'],[50, 'D'],[33, 'F'],[63, 'C'],
     [48, 'F'],[30, 'F'],[54, 'D'],[30, 'F'],[39, 'F'],[42, 'F'],[59, 'D+'],
     [58, 'D+'],[61, 'C'],[24, 'F'],[28, 'F'],[58, 'D+'],[46, 'F'],[66, 'C+'],
     [54, 'D'],[64, 'C'],[79, 'B+'],[33, 'F'],[44, 'F'],[60, 'C'],[15, 'F'],
     [48, 'F'],[62, 'C'],[57, 'D+'],[46, 'F'],[51, 'D'],[50, 'D'],[27, 'F'],
     [69, 'C+'],[59, 'D+'],[62, 'C'],[24, 'F'],[55, 'D+'],[58, 'D+'],[51,'D'],
     [28, 'F'],[65, 'C+'],[47, 'F'],[11, 'F'],[90, 'A'],[69, 'C+'],[43, 'F'],
     [75, 'B+'],[53, 'D'],[40, 'F'],[42, 'F'],[53, 'D'],[61, 'C'],[35, 'F'],
     [34, 'F'],[52, 'D'],[52, 'D'],[55, 'D+'],[36, 'F'],[36, 'F'],[79, 'B+'],
     [46, 'F'],[88, 'A'],[78, 'B+'],[18, 'F'],[87, 'A'],[46, 'F'],[30, 'F'],
     [36, 'F'],[51, 'D'],[59, 'D+'],[16, 'F'],[64, 'C'],[54, 'D'],[42, 'F'],
     [50, 'D'],[51, 'D'],[40, 'F'],[44, 'F'],[61, 'C'],[35, 'F'],[52, 'D'],
     [15, 'F'],[68, 'C+']],
     columns=['Total (100%)', 'Grade']
)

In [31]:
DF

Total (100%) Grade
0             62     C
1             77    B+
2             27     F
3             44     F
4             58    D+
..           ...   ...
95            61     C
96            35     F
97            52     D
98            15     F
99            68    C+

[100 rows x 2 columns]

In [32]:
DF[["Grade","Total (100%)"]]

Grade  Total (100%)
0      C            62
1     B+            77
2      F            27
3      F            44
4     D+            58
..   ...           ...
95     C            61
96     F            35
97     D            52
98     F            15
99    C+            68

[100 rows x 2 columns]

Let's play some stupid stuff 🤪

In [33]:
#You able to recall the repetive column

DF[["Grade","Total (100%)","Grade"]]

Grade  Total (100%) Grade
0      C            62     C
1     B+            77    B+
2      F            27     F
3      F            44     F
4     D+            58    D+
..   ...           ...   ...
95     C            61     C
96     F            35     F
97     D            52     D
98     F            15     F
99    C+            68    C+

[100 rows x 3 columns]

In [34]:
My_index =[90,80,70,90,80,70,90,80,70]

In [35]:
#You able to run the same row

DF.loc[My_index]

Total (100%) Grade
90            42     F
80            18     F
70            34     F
90            42     F
80            18     F
70            34     F
90            42     F
80            18     F
70            34     F



---



### 1.2 ) The characteristic of the column(s)

In [36]:
DF.describe()

Total (100%)
count    100.000000
mean      49.860000
std       17.141137
min       11.000000
25%       37.750000
50%       51.000000
75%       61.000000
max       90.000000

In [37]:
# The characteristic of the first column (according to all rows)

DF.iloc[:,0].describe()

count    100.000000
mean      49.860000
std       17.141137
min       11.000000
25%       37.750000
50%       51.000000
75%       61.000000
max       90.000000
Name: Total (100%), dtype: float64

In [38]:
# Characteristic of Column Grade 

DF.iloc[:,1].describe()

count     100
unique      8
top         F
freq       45
Name: Grade, dtype: object

In [39]:
# Characteristic of Column Grade (version2)

DF.Grade.describe()

count     100
unique      8
top         F
freq       45
Name: Grade, dtype: object

In [40]:
X = DF.Grade.value_counts()

print("Now X is ",type(X))        #Pandas Series
X

Now X is  <class 'pandas.core.series.Series'>


F     45
D     17
D+    11
C     10
B+     7
C+     6
A      3
B      1
Name: Grade, dtype: int64



---


## <u>Application 2</u> : Student Grade but larger imported dataset

### 2.1 ) Import Dataset

- With Index column

In [41]:
url = 'https://raw.githubusercontent.com/wallik2/SCIM_Lecture/main/Intro_Data_Sci/Raw_Score.csv'

df = pd.read_csv(url)

In [42]:
df.head(5) 

ID  Assignments (100)  Midterm (100)  ...  Final (35%)  Total (100%)  Grade
0  XX01001                 86             56  ...          NaN           NaN    NaN
1  XX01002                 16             93  ...          NaN           NaN    NaN
2  XX01003                 50             46  ...          NaN           NaN    NaN
3  XX01004                 96             36  ...          NaN           NaN    NaN
4  XX01005                 13             83  ...          NaN           NaN    NaN

[5 rows x 9 columns]

- Without Index column

In [43]:
url = 'https://raw.githubusercontent.com/wallik2/SCIM_Lecture/main/Intro_Data_Sci/Raw_Score.csv'

df = pd.read_csv(url,index_col=0)

In [44]:
df.head(5)

Assignments (100)  Midterm (100)  ...  Total (100%)  Grade
ID                                         ...                     
XX01001                 86             56  ...           NaN    NaN
XX01002                 16             93  ...           NaN    NaN
XX01003                 50             46  ...           NaN    NaN
XX01004                 96             36  ...           NaN    NaN
XX01005                 13             83  ...           NaN    NaN

[5 rows x 8 columns]

df.columns



---


### 2.2) Characteristics of DataFrame

In [45]:
df.columns

Index(['Assignments (100)', 'Midterm (100)', 'Final (100)',
       'Assignments (30%)', 'Midterm (35%)', 'Final (35%)', 'Total (100%)',
       'Grade'],
      dtype='object')

> Ways to access the first column

In [46]:
df['Assignments (100)']

ID
XX01001     86
XX01002     16
XX01003     50
XX01004     96
XX01005     13
          ... 
XX03998     56
XX03999    100
XX04001     90
XX04002     26
XX04003     33
Name: Assignments (100), Length: 3000, dtype: int64

In [47]:
df.iloc[:,0]

#แบนี้ ตารางจะไม่สวย

ID
XX01001     86
XX01002     16
XX01003     50
XX01004     96
XX01005     13
          ... 
XX03998     56
XX03999    100
XX04001     90
XX04002     26
XX04003     33
Name: Assignments (100), Length: 3000, dtype: int64

In [48]:
df.iloc[:,[0]]

#ใส่ list ให้ 0 เพิ้อให้ ได้ตารางแบบนี้

Assignments (100)
ID                        
XX01001                 86
XX01002                 16
XX01003                 50
XX01004                 96
XX01005                 13
...                    ...
XX03998                 56
XX03999                100
XX04001                 90
XX04002                 26
XX04003                 33

[3000 rows x 1 columns]

In [49]:
df.loc[:,['Assignments (100)']]

Assignments (100)
ID                        
XX01001                 86
XX01002                 16
XX01003                 50
XX01004                 96
XX01005                 13
...                    ...
XX03998                 56
XX03999                100
XX04001                 90
XX04002                 26
XX04003                 33

[3000 rows x 1 columns]

In [50]:
df.iloc[:,0:1]

Assignments (100)
ID                        
XX01001                 86
XX01002                 16
XX01003                 50
XX01004                 96
XX01005                 13
...                    ...
XX03998                 56
XX03999                100
XX04001                 90
XX04002                 26
XX04003                 33

[3000 rows x 1 columns]

> Ways to access the first row

In [51]:
df.iloc[0]

# เวอร์ชั่นไม่สวย

Assignments (100)    86.0
Midterm (100)        56.0
Final (100)          60.0
Assignments (30%)     NaN
Midterm (35%)         NaN
Final (35%)           NaN
Total (100%)          NaN
Grade                 NaN
Name: XX01001, dtype: float64

In [52]:
df.iloc[[0]]

# เวอร์ชั่นสวย

Assignments (100)  Midterm (100)  ...  Total (100%)  Grade
ID                                         ...                     
XX01001                 86             56  ...           NaN    NaN

[1 rows x 8 columns]

In [53]:
df.iloc[0:4,0:4]

Assignments (100)  Midterm (100)  Final (100)  Assignments (30%)
ID                                                                       
XX01001                 86             56           60                NaN
XX01002                 16             93           80                NaN
XX01003                 50             46           40                NaN
XX01004                 96             36           62                NaN

> Ways to access multiple rows

In [54]:
#Alternative 1
df.iloc[0:10]

Assignments (100)  Midterm (100)  ...  Total (100%)  Grade
ID                                         ...                     
XX01001                 86             56  ...           NaN    NaN
XX01002                 16             93  ...           NaN    NaN
XX01003                 50             46  ...           NaN    NaN
XX01004                 96             36  ...           NaN    NaN
XX01005                 13             83  ...           NaN    NaN
XX01006                 70              3  ...           NaN    NaN
XX01007                 86             60  ...           NaN    NaN
XX01008                 33             36  ...           NaN    NaN
XX01009                 26             76  ...           NaN    NaN
XX01010                 26              3  ...           NaN    NaN

[10 rows x 8 columns]

In [55]:
#Alternative 2

df.head(10)

Assignments (100)  Midterm (100)  ...  Total (100%)  Grade
ID                                         ...                     
XX01001                 86             56  ...           NaN    NaN
XX01002                 16             93  ...           NaN    NaN
XX01003                 50             46  ...           NaN    NaN
XX01004                 96             36  ...           NaN    NaN
XX01005                 13             83  ...           NaN    NaN
XX01006                 70              3  ...           NaN    NaN
XX01007                 86             60  ...           NaN    NaN
XX01008                 33             36  ...           NaN    NaN
XX01009                 26             76  ...           NaN    NaN
XX01010                 26              3  ...           NaN    NaN

[10 rows x 8 columns]



---

### TIPS: How to Check the CPU time

In [29]:
import time
T1 = time.time()

# Your process starts here
for i in range(0,20000000):
  i = i+2
# End of process

T2 = time.time()

print("CPU Time is", T2-T1)

CPU Time is 1.5499207973480225


when I use nothing : 1.811515153215 seconds<br>
when I use GPU : 1.5499207973480225 seconds<br>
when I use TPU: CPU Time is 2.274764060974121 seconds



---


TIPS2: Chop or Round number

In [39]:

A = 12.9
print("int(", A ,") \t is ",int(A))
print("round(", A ,") is ",round(A))

int( 12.9 ) 	 is  12
round( 12.9 ) is  13




---
###2.3) Let fill the blank cell with the computed grade


In [41]:
def Grade(Total):

#for i in range(0,len(df)):
  if Total >= 80:
    return("A")
  elif Total >= 75:
    return("B+")
  elif Total >= 70:
    return("B")
  elif Total >= 65:
    return("C+")
  elif Total >= 60:
    return("C")
  elif Total >= 55:
    return("D+")
  elif Total >= 50:
    return("D")
  else: return("F")

In [42]:
Grade(79)

'B+'